pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/SumWithPrecisionAggregationFunction.java

```diff
@@ -0,0 +1,174 @@
+      groupByResultValue = groupByResultValue.add(value, _mathContext);
+      groupByResultHolder.setValueForKey(groupKey, setScale(groupByResultValue));
+    }
+  }
+
+  @Override
+  public void aggregateGroupByMV(int length, int[][] groupKeysArray, GroupByResultHolder groupByResultHolder,
+      Map<ExpressionContext, BlockValSet> blockValSetMap) {
+    byte[][] valueArray = blockValSetMap.get(_expression).getBytesValuesSV();
+    for (int i = 0; i < length; i++) {
+      byte[] value = valueArray[i];
+      for (int groupKey : groupKeysArray[i]) {
+        BigDecimal groupByResultValue = getDefaultResult(groupByResultHolder, groupKey);
+        BigDecimal valueBigDecimal = new BigDecimal(DataTypeConversionFunctions.bytesToBigDecimal(value));
+        groupByResultValue = groupByResultValue.add(valueBigDecimal, _mathContext);
+        groupByResultHolder.setValueForKey(groupKey, setScale(groupByResultValue));
+      }
+    }
+  }
+
+  @Override
+  public BigDecimal extractAggregationResult(AggregationResultHolder aggregationResultHolder) {
+    return getDefaultResult(aggregationResultHolder);
+  }
+
+  @Override
+  public BigDecimal extractGroupByResult(GroupByResultHolder groupByResultHolder, int groupKey) {
+    return getDefaultResult(groupByResultHolder, groupKey);
+  }
+
+  @Override
+  public BigDecimal merge(BigDecimal intermediateResult1, BigDecimal intermediateResult2) {
+    try {
+      return setScale(intermediateResult1.add(intermediateResult2, _mathContext));
+    } catch (Exception e) {
+      throw new RuntimeException("Caught Exception while merging results in sum with precision function", e);
+    }
+  }
+
+  @Override
+  public boolean isIntermediateResultComparable() {
+    return true;
+  }
+
+  @Override
+  public DataSchema.ColumnDataType getIntermediateResultColumnType() {
+    return DataSchema.ColumnDataType.OBJECT;
+  }
+
+  @Override
+  public DataSchema.ColumnDataType getFinalResultColumnType() {
+    return DataSchema.ColumnDataType.STRING;
+  }
+
+  @Override
+  public BigDecimal extractFinalResult(BigDecimal intermediateResult) {
+    return intermediateResult;
+  }
+
+  public BigDecimal getDefaultResult(AggregationResultHolder aggregationResultHolder) {
+    BigDecimal result = aggregationResultHolder.getResult();
+    if (result == null) {
+      result = new BigDecimal(0, _mathContext);
+      aggregationResultHolder.setValue(result);
+    }
+    result = setScale(result);
+    return result;
+  }
+
+  public BigDecimal getDefaultResult(GroupByResultHolder groupByResultHolder, int groupKey) {
+    BigDecimal result = groupByResultHolder.getResult(groupKey);
+    if (result == null) {
+      result = new BigDecimal(0, _mathContext);
+      groupByResultHolder.setValueForKey(groupKey, result);
+    }
+    result = setScale(result);
+    return result;
+  }
+
+  private BigDecimal setScale(BigDecimal value) {
```

Comment:
this method is bit confusing, what is it trying to do. add some javadocs and maybe rename the function if possible

Reply:
If the user doesn't set scale, then just return the value as such otherwise set the scale value.
Just a small helper method to avoid if else in the code.

Reply:
got it. I am assuming big decimal additions are idempotent -- bd1 + bd2 same as bd2 + bd1